File: src/main/org/nlogo/prim/threed/_patchat.java
package org.nlogo.prim.threed;

import org.nlogo.agent.Patch;
import org.nlogo.api.LogoException;
import org.nlogo.nvm.Context;
import org.nlogo.nvm.Reporter;
import org.nlogo.nvm.Syntax;

public final strictfp class _patchat
    extends Reporter {
  @Override
  public Syntax syntax() {
    int[] right = {Syntax.TYPE_NUMBER,
        Syntax.TYPE_NUMBER,
        Syntax.TYPE_NUMBER};
    int ret = Syntax.TYPE_PATCH;
    return Syntax.reporterSyntax(right, ret, "-TP-");
  }

  @Override
  public Object report(final Context context) throws LogoException {
    double dx = argEvalDoubleValue(context, 0);
    double dy = argEvalDoubleValue(context, 1);
    double dz = argEvalDoubleValue(context, 2);
    Patch patch = null;
    try {
      patch = ((org.nlogo.agent.Agent3D) (context.agent)).getPatchAtOffsets(dx, dy, dz);
    } catch (org.nlogo.api.AgentException e) {
      return org.nlogo.api.Nobody$.MODULE$;
    }
    return patch;
  }
}

File: src/main/org/nlogo/prim/threed/_patchat.scala
package org.nlogo.prim.threed

import org.nlogo.agent.{ Agent3D, Patch }
import org.nlogo.api.{ AgentException, Nobody }
import org.nlogo.nvm.{ Context, Reporter, Syntax }

class _patchat extends Reporter {
  override def syntax =
    Syntax.reporterSyntax(Array(Syntax.TYPE_NUMBER,
                                Syntax.TYPE_NUMBER,
                                Syntax.TYPE_NUMBER),
                          Syntax.TYPE_PATCH, "-TP-")
  override def report(context: Context) = {
    try context.agent.asInstanceOf[Agent3D].getPatchAtOffsets(
      argEvalDoubleValue(context, 0),
      argEvalDoubleValue(context, 1),
      argEvalDoubleValue(context, 2))
    catch { case _: AgentException => Nobody }
  }
}
